Warn when targets relabelled to same labels

darshanime opened the pull request:
closes https://github.com/prometheus/prometheus/issues/5136
Signed-off-by: darshanime <deathbullet@gmail.com>

<!--
    Don't forget!

    - Please sign CNCF's Developer Certificate of Origin and sign-off your commits by adding the -s / --sign-off flag to `git commit`. See https://github.com/apps/dco for more information.

    - If the PR adds or changes a behaviour or fixes a bug of an exported API it would need a unit/e2e test.

    - Where possible use only exported APIs for tests to simplify the review and make it as close as possible to an actual library usage.

    - No tests are needed for internal implementation changes.

    - Performance improvements would need a benchmark test to prove it.

    - All exposed objects should have a comment.

    - All comments should start with a capital letter and end with a full stop.
 -->

LeviHarrison reviewed (commented):
Thanks! Just a few comments.

darshanime commented:
thanks @LeviHarrison, addressed your comments. 

> This indeed does not belong to the discovery manager but to the scrape manager.

@roidelapluie the logic is currently in scrape manager's `reload()` method, after the `Sync()` has been called for each group. Do you think it should be someplace else?

darshanime commented on scrape/manager.go:
```diff
@@ -209,6 +211,17 @@ func (m *Manager) reload() {
+	m.activeTargets = make(map[string]*Target)
+	for _, scrapePool := range m.scrapePools {
+		for _, target := range scrapePool.activeTargets {
+			lset := relabel.Process(target.Labels(), scrapePool.config.MetricRelabelConfigs...)
```
dang, thanks!

LeviHarrison commented on scrape/manager.go:
```diff
@@ -205,10 +205,19 @@ func (m *Manager) reload() {
+	for _, scrapePool := range m.scrapePools {
+		for _, target := range scrapePool.activeTargets {
+			if t, ok := activeTargets[target.hash()]; ok {
+				level.Warn(m.logger).Log("msg", "Found targets with same labels after relabelling", "target", t, "target", target)
```
The issue here is that the identifiers for the targets (`t` and `target`) are a 64-bit integer and a URL, the former being not very helpful, and both not matching.

I'm partial to just including the URL of the target, and because both will be the same we only need one, but I'll turn to @roidelapluie for a second opinion.

LeviHarrison replied:
Actually, my comment is wrong. Both of these come out to be the URL of the target, and since that's the same, we only need one.

LeviHarrison commented:
> @roidelapluie the logic is currently in scrape manager's reload() method, after the Sync() has been called for each group. Do you think it should be someplace else?

I think @roidelapluie is affirming the location in this PR is correct.

LeviHarrison commented:
Could you please also add a quick test for this?

beorn7 commented:
Picking this up during our bug scrub. @darshanime are you still up to adding a test?

darshanime committed:
Add warning log for same labelset after relabeling

Signed-off-by: darshanime <deathbullet@gmail.com>

darshanime committed:
Add test for same label set in targets

Signed-off-by: darshanime <deathbullet@gmail.com>

bboreham commented on scrape/manager.go:
```diff
@@ -184,6 +185,20 @@ func (m *Manager) reload() {
+		for _, target := range scrapePool.activeTargets {
+			if t, ok := activeTargets[target.labels.Hash()]; ok {
+				level.Warn(m.logger).Log(
+					"msg", "Found targets with same labels after relabelling",
```
I wonder if this should print the full set of labels?  (`target.labels.String()`)

bboreham reviewed (commented):
Could extract the new check to its own function - `reload` is getting a bit long.

darshanime committed:
Extract to separate method

Signed-off-by: darshanime <deathbullet@gmail.com>

GiedriusS commented on scrape/manager.go:
```diff
@@ -184,6 +185,25 @@ func (m *Manager) reload() {
+	activeTargets := make(map[uint64]*Target)
+	for _, scrapePool := range m.scrapePools {
+		for _, target := range scrapePool.activeTargets {
+			if t, ok := activeTargets[target.labels.Hash()]; ok {
```
Nit: I think we should return early.
```
t, ok := activeTargets[target.labels.Hash()]
if !ok {
  continue
}

... log here ...
```

darshanime replied:
updated, TFR!

machine424 commented:
How about doing this here https://github.com/prometheus/prometheus/blob/eea6ab1cdd24ec69c94ba4b0d165030c89860c8b/scrape/scrape.go#L486 instead of re-looping again?

We may need to do the same for `notifier/notifier.go`, I don't know if it's possible to get sups in there.

darshanime committed:
Invert predicate to use !ok

Signed-off-by: darshanime <deathbullet@gmail.com>

darshanime commented:
ran the benchmark:
```
$ go test -bench=BenchmarkScrapeLoop -run=- -count 6 | tee main
$ go test -bench=BenchmarkScrapeLoop -run=- -count 6 | tee duplicate_targets
$ benchstat main duplicate_targets
goos: darwin
goarch: arm64
pkg: github.com/prometheus/prometheus/scrape
                      │    main     │         duplicate_targets         │
                      │   sec/op    │   sec/op     vs base              │
ScrapeLoopAppend-10     37.58µ ± 2%   37.31µ ± 1%       ~ (p=0.065 n=6)
ScrapeLoopAppendOM-10   36.76µ ± 1%   36.64µ ± 1%       ~ (p=0.485 n=6)
geomean                 37.16µ        36.97µ       -0.51%
```

bboreham commented:
Please add `-benchmem` to the benchmark.
Also check that benchmark calls the function you are changing.

darshanime committed:
Add benchmark for warnIfTargetsRelabelledToSameLabels

Signed-off-by: darshanime <deathbullet@gmail.com>

darshanime committed:
Preallocate map with required size

Signed-off-by: darshanime <deathbullet@gmail.com>

darshanime commented:
@bboreham, i have created a new benchmark; as expected the operation isn't memory intensive for 10k targets...

```
$ go test -bench=BenchmarkManagerReload -benchmem -run=- -count 6  -benchtime=10000x
goos: darwin
goarch: arm64
pkg: github.com/prometheus/prometheus/scrape
BenchmarkManagerReload-10    	   10000	    703110 ns/op	  319529 B/op	       3 allocs/op
BenchmarkManagerReload-10    	   10000	    698015 ns/op	  319529 B/op	       3 allocs/op
BenchmarkManagerReload-10    	   10000	    706037 ns/op	  319529 B/op	       3 allocs/op
BenchmarkManagerReload-10    	   10000	    701210 ns/op	  319529 B/op	       3 allocs/op
BenchmarkManagerReload-10    	   10000	    700095 ns/op	  319529 B/op	       3 allocs/op
BenchmarkManagerReload-10    	   10000	    695985 ns/op	  319529 B/op	       3 allocs/op
PASS
ok  	github.com/prometheus/prometheus/scrape	42.650s
```

bboreham commented on discovery/targetgroup/targetgroup.go:
```diff
@@ -20,7 +20,7 @@ import (
 )
 
-// Group is a set of targets with a common label set(production , test, staging etc.).
+// Group is a set of targets with a common label set(production, test, staging etc).
```
Change seems unrelated.

darshanime replied:
Yes, unrelated but entirely routine and uncontroversial. Are you suggesting I remove it?

bboreham replied:
Yes.  When I look back over the history of a file I want to see changes labeled with the reason they were made.
For me it is routine and uncontroversial to put distinct changes in different PRs.

darshanime replied:
Okay, fair enough. Will also remove the new lines I added elsewhere in the PR.

bboreham commented on scrape/manager_test.go:
```diff
@@ -610,6 +610,64 @@ func TestManagerTargetsUpdates(t *testing.T) {
+		activeTargets: map[uint64]*Target{},
+	}
+
+	for i := 0; i < b.N; i++ {
```
It is essential to do the same amount of work each time the benchmark is run, so varying the number of targets with `b.N` is wrong.

darshanime replied:
We are running the `reload` function 10k times, each time with 10k targets via `go test -bench=BenchmarkManagerReload -benchmem -run=- -count 6  -benchtime=10000x`. This is similar to this [benchmark](https://github.com/prometheus/prometheus/blob/main/tsdb/chunkenc/chunk_test.go#L266). 

Do you think it would be better to hardcode the #targets 10k instead?

bboreham replied:
> This is similar to this [benchmark](https://github.com/prometheus/prometheus/blob/main/tsdb/chunkenc/chunk_test.go#L266).

Another invalid benchmark.

> Do you think it would be better to hardcode the #targets 10k instead?

Yes.

bboreham commented on scrape/manager_test.go:
```diff
@@ -610,6 +610,64 @@ func TestManagerTargetsUpdates(t *testing.T) {
+	m.scrapePools["default"] = sp
+
+	m.reload()
+	require.Contains(t, output, "Found targets with same labels after relabelling")
```
Does this test do any relabeling?

darshanime replied:
Nope, we manually add 2 targets with the same label sets and assert that the log output contains the desired warning.

bboreham commented on scrape/manager.go:
```diff
@@ -190,6 +192,36 @@ func (m *Manager) reload() {
+	activeTargets := make(map[uint64]*Target, totalTargets)
+	for _, scrapePool := range m.scrapePools {
+		for _, target := range scrapePool.activeTargets {
+			lHash := target.labels.Hash()
+			t, ok := activeTargets[lHash]
```
There is some risk that two sets of labels will hash to the same value; it would be safer to make the map key `labels.Bytes()`.

bboreham reviewed (commented):
When posting benchmark results it is traditional to give the before/after comparison.
However when I tried to run the benchmark against the code "before", it hung.
This is because the benchmark does more work when it runs faster, which makes it an invalid benchmark.

darshanime commented:
> When posting benchmark results it is traditional to give the before/after comparison.
However when I tried to run the benchmark against the code "before", it hung.
This is because the benchmark does more work when it runs faster, which makes it an invalid benchmark.

You are right about the delta being the traditional way to showcase benchmark results, but (as you found out too), without this patch, the loop does so little work that the numbers don't register at all (are mostly all 0s). How do you wish to proceed from here? imo, the "benchmark" shows that the patch only adds a single, inexpensive pass thru the target set; so it did its work.

darshanime committed:
Use label bytes as key

Signed-off-by: darshanime <deathbullet@gmail.com>

bboreham commented:
>  How do you wish to proceed from here?

Write a valid benchmark, that does the same amount of work per iteration.

darshanime committed:
Fix the target count to 10k

Signed-off-by: darshanime <deathbullet@gmail.com>

darshanime committed:
Revert unrelated changes

Signed-off-by: darshanime <deathbullet@gmail.com>

bboreham commented:
> the benchmark without this patch is not interesting

I think it's interesting to know what the delta is. About 0.8ms for 10K targets.  On a large Kubernetes cluster, with changes happening all the time, that might be significant.  I'm not sure what the true baseline would be.

20ns is sufficiently small that it makes me check what the benchmark does in the 'before' version, which is nothing at all.
`m.targetSets` is not initialized, so that loop falls through.  So it's not a benchmark of the `reload` function, and should not be named as if it is.

bboreham commented:
Hello from the bug-scrub!

We wondered whether this requirement is now covered by #12933?

If not, we noted that the question of overhead could be avoided by putting this behind a feature flag.

darshanime committed:
add warn under feature flag

Signed-off-by: darshanime <deathbullet@gmail.com>

beorn7 commented:
Hello again from the bug-scrub! 

@darshanime thanks for further work on this, and our apologies for not coming back to this sooner. (It would have helped if you stated that this is ready for another round of review. Which we assume is the case.)

This is now behind a feature flag (maybe the name should be a bit shorter?). It also appears clear to us that https://github.com/prometheus/prometheus/pull/12933 has addressed a sufficiently different problem and this is still worth having.

@machine424 will continue the review here. (@bboreham if you still have mental context on this, your help would also be appriciated.)

machine424 reviewed (commented):
Thanks for working on this.
It just needs a few final touches and we can merge it.

darshanime committed:
address pr comments

Signed-off-by: darshanime <deathbullet@gmail.com>

darshanime commented on scrape/manager.go:
```diff
@@ -205,6 +208,40 @@ func (m *Manager) reload() {
+	builder := labels.NewBuilder(labels.EmptyLabels())
+	for _, scrapePool := range m.scrapePools {
+		for _, target := range scrapePool.activeTargets {
+			lStr := string(target.labels.Bytes(buf[:]))
```
reverted to using `hash()`; seems safe enough since that is used for the [sync loop](https://github.com/prometheus/prometheus/blob/v3.4.0-rc.0/scrape/scrape.go#L531) as well

darshanime commented on scrape/manager_test.go:
```diff
@@ -622,6 +622,43 @@ func TestManagerTargetsUpdates(t *testing.T) {
+	}
+	sp.activeTargets[uint64(0)] = &Target{
+		scrapeConfig: targetScrapeCfg,
+		tLabels:      map[model.LabelName]model.LabelValue{model.AddressLabel: "foo"},
```
updated the test to make it clearer

machine424 commented on scrape/manager_test.go:
```diff
@@ -616,6 +616,42 @@ func TestManagerTargetsUpdates(t *testing.T) {
+		ScrapeTimeout:  model.Duration(time.Second),
+	}
+
+	sp.activeTargets[uint64(0)] = &Target{
```
I feel this is too low‑level and too tightly coupled to the implementation, how about passing a config like in https://github.com/prometheus/prometheus/blob/091e662f4d2e7b0e0310844fe49d7297d17f1bc4/scrape/scrape_test.go#L5173-L5204, trigger and reload and wait for the warning?

machine424 commented on scrape/manager.go:
```diff
@@ -208,6 +211,39 @@ func (m *Manager) reload() {
+			}
+			m.logger.Warn(
+				"Found targets with same labels after relabelling",
+				"target_one", t.DiscoveredLabels(builder).Get(model.AddressLabel),
```
if the labels are the same, the address label value should be identical, unless there is a hash collision.
logging the labels is enough.

Also, in `docs/feature_flags.md` we may want to explicitly say that the log will contain the labels, in case someone is storing some sensitive data in there.

machine424 reviewed (commented):
We’ve never been closer!
Some propositions.

darshanime commented on scrape/manager.go:
```diff
@@ -208,6 +211,39 @@ func (m *Manager) reload() {
+	defer m.mtxScrape.Unlock()
+
+	totalTargets := 0
+	for _, scrapePool := range m.scrapePools {
```
yup, i ran the benchmark and using this loop to get the size upfront was beneficial overall

darshanime committed:
update test

Signed-off-by: darshanime <deathbullet@gmail.com>

darshanime committed:
add feature flag to list

Signed-off-by: darshanime <deathbullet@gmail.com>

darshanime committed:
make cli-documentation

Signed-off-by: darshanime <deathbullet@gmail.com>

machine424 commented on scrape/manager_test.go:
```diff
@@ -616,6 +617,75 @@ func TestManagerTargetsUpdates(t *testing.T) {
 	}
 }
 
+func TestManagerDuplicateAfterRelabellingWarning(t *testing.T) {
```
the test is still doing too many things/implementing some logic that we want to test.

 let's go with a higher level test like in https://github.com/machine424/prometheus/commit/678b94651fe8329b776443574295d524c96cd2ee, you can cherry-pick the whole commit.
 
 please double check as the test is mainly generated by an LLM (Claude Sonnet 4)

darshanime replied:
i cherry-picked the commit, but the test therein takes `~5s` to run, even after passing `"--scrape.discovery-reload-interval=500ms"` as an extra arg. on the other hand, in the test currently implemented, since we call `m.reload()` explicitly and get it finished in ~350ms reliably. 

i have refactored the test, can you check if it looks better now? else, i can cherry-pick 👍

darshanime committed:
refactor test to remove redundant assertions

Signed-off-by: darshanime <deathbullet@gmail.com>

bboreham commented:
Hello from the bug-scrub!  @machine424 will take another look.

bboreham commented on scrape/manager.go:
```diff
@@ -213,6 +216,36 @@ func (m *Manager) reload() {
+	for _, scrapePool := range m.scrapePools {
+		for _, target := range scrapePool.activeTargets {
+			tHash := target.hash()
+			_, ok := activeTargets[tHash]
```
We would need to add this to the list of places where hash collision is ignored.
* https://github.com/prometheus/prometheus/issues/12519#issuecomment-1776886428

Or change the code to check for exact match.

bboreham commented:
Closing due to lack of response.  Happy to re-open if you come back to it @darshanime.
